File: implementations/go/eapp/discovery.go
package eapp

import (
	"sort"
	"strings"
)

// Criteria filters discovery results (§8.1).
//
// Every field is optional; an empty Criteria matches every plugin in scope.
type Criteria struct {
	// Capability is the capability name to match (exact, case-sensitive).
	Capability string `json:"capability,omitempty"`
	// Version is a SemVer *range*, not a version (§8.1). An exact version is
	// also a valid range, so both readings work; an unsupported range is
	// rejected rather than silently matching nothing (errata E-E).
	Version string `json:"version,omitempty"`
	// Constraints are matched exactly against the offered capability's
	// constraints (C-7). All of them must match — see constraintSetMatches.
	Constraints []Constraint `json:"constraints,omitempty"`
	// Identity is a partial identity filter. An absent field means "any"; an
	// empty string is treated the same way, because the wire form cannot
	// distinguish an omitted member from an empty one for a string field and
	// guessing otherwise would make the filter non-portable.
	Identity PartialIdentity `json:"identity,omitempty"`
}

// PartialIdentity is the `Partial<Identity>` of §8.1.
//
// It is a distinct type from Identity on purpose: Identity is the frozen
// three-field record whose completeness is checked by ID-1…ID-3, whereas this
// is a query that is *allowed* to be incomplete. Reusing Identity would mean
// every discovery query had to invent two of its three fields.
type PartialIdentity struct {
	Domain   string `json:"domain,omitempty"`
	ID       string `json:"id,omitempty"`
	Instance string `json:"instance,omitempty"`
}

// Matches reports whether a full identity satisfies the partial filter.
func (p PartialIdentity) Matches(id Identity) bool {
	if p.Domain != "" && p.Domain != id.Domain {
		return false
	}
	if p.ID != "" && p.ID != id.ID {
		return false
	}
	if p.Instance != "" && p.Instance != id.Instance {
		return false
	}
	return true
}

// IsZero reports whether the filter constrains nothing.
func (p PartialIdentity) IsZero() bool {
	return p.Domain == "" && p.ID == "" && p.Instance == ""
}

// DiscoveryScope narrows a discovery query (§8.1).
//
// D-7 is the invariant that matters here: trust levels are a *classification*
// (L0 / L1 / L2), NOT an ordered authorization ladder. Nothing in this package
// compares two levels with < or >, and no code path derives "L2 may see L1"
// from the level values — visibility comes exclusively from the injected
// VisibilityPolicy, so an implementation cannot accidentally turn a taxonomy
// into a ranking.
type DiscoveryScope struct {
	// TrustLevel is one of L0, L1, L2; empty means "any".
	TrustLevel string `json:"trustLevel,omitempty"`
	// TrustDomain is a deployment/trust domain string; empty means "any".
	TrustDomain string `json:"trustDomain,omitempty"`
}

// Trust level constants. They are opaque labels: the ordering of these
// declarations carries no meaning.
const (
	TrustLevelL0 = "L0"
	TrustLevelL1 = "L1"
	TrustLevelL2 = "L2"
)

// Valid reports whether the scope is well formed (finite enum + string domain).
func (s DiscoveryScope) Valid() error {
	switch s.TrustLevel {
	case "", TrustLevelL0, TrustLevelL1, TrustLevelL2:
		return nil
	default:
		return errDiscoveryScopeInvalid("trustLevel %q MUST be one of L0, L1, L2 (D-7)", s.TrustLevel)
	}
}

// String renders the scope for diagnostics.
func (s DiscoveryScope) String() string {
	level := s.TrustLevel
	if level == "" {
		level = "*"
	}
	domain := s.TrustDomain
	if domain == "" {
		domain = "*"
	}
	return level + "/" + domain
}

// VisibilityPolicy answers "may a plugin be seen through this scope?".
//
// The Composition Core does not put trust metadata on a Plugin: §5.1 freezes
// Plugin to {identity, capabilities, lifecycle} and §8.2 says trust level is a
// deployment classification, not a plugin attribute. So the Core cannot compute
// visibility on its own without inventing a field the spec does not define.
// Injecting the policy keeps D-1/D-2 enforceable (find and watch consult the
// same function) while leaving the metadata question to the deployment, which
// is where the spec leaves it.
type VisibilityPolicy interface {
	// Visible reports whether plugin is visible through scope.
	Visible(plugin Plugin, scope DiscoveryScope) bool
}

// VisibilityFunc adapts a function to VisibilityPolicy.
type VisibilityFunc func(plugin Plugin, scope DiscoveryScope) bool

// Visible implements VisibilityPolicy.
func (f VisibilityFunc) Visible(plugin Plugin, scope DiscoveryScope) bool { return f(plugin, scope) }

// openVisibility is the default policy: every plugin is visible in every scope.
//
// It is the only default that does not fabricate a rule. A Core with no trust
// metadata will report every plugin for any scope — including a scope naming a
// trust level that nothing is tagged with — which is honest ("this deployment
// declares no trust boundaries") and lets D-1/D-2 be tested by injecting a
// policy that does restrict things.
func openVisibility() VisibilityPolicy {
	return VisibilityFunc(func(Plugin, DiscoveryScope) bool { return true })
}

// EventType is the frozen DiscoveryEvent.type of D-6.
type EventType string

const (
	// EventAdded fires when a plugin becomes visible in a scope.
	EventAdded EventType = "added"
	// EventRemoved fires when a plugin stops being visible in a scope.
	EventRemoved EventType = "removed"
	// EventChanged fires when a visible plugin's capabilities change (P-4).
	EventChanged EventType = "changed"
)

// Valid reports whether t is one of added / removed / changed (D-6).
func (t EventType) Valid() bool {
	switch t {
	case EventAdded, EventRemoved, EventChanged:
		return true
	default:
		return false
	}
}

// DiscoveryEvent is what a watcher receives (§8.1).
type DiscoveryEvent struct {
	Type   EventType `json:"type"`
	Plugin Identity  `json:"plugin"`
}

// matchesCriteria reports whether a plugin satisfies a criteria set.
//
// Semantics worth stating:
//   - A capability name is required for the capability/version/constraints
//     clauses to mean anything; without it, a version filter would have to
//     apply "some capability", which makes the result order-dependent and
//     therefore unstable.
//   - Any one capability may satisfy the whole clause. A plugin that exposes
//     `casing.apply@1.0.0` and `casing.apply@2.0.0` and is queried for
//     `version:"^1.0.0"` matches through the first, and MUST be reported once.
func matchesCriteria(plugin Plugin, criteria Criteria, rangeCache map[string]*SemverRange) (bool, error) {
	if !criteria.Identity.Matches(plugin.Identity) {
		return false, nil
	}
	if criteria.Capability == "" {
		// No capability clause: constraints/version cannot be evaluated against
		// a specific capability, so they are ignored rather than guessed at.
		return true, nil
	}

	var versionRange *SemverRange
	if strings.TrimSpace(criteria.Version) != "" {
		cached, ok := rangeCache[criteria.Version]
		if !ok {
			parsed, err := ParseRange(criteria.Version)
			if err != nil {
				// Explicit rejection, never a silent empty result (errata E-E).
				return false, errDiscoveryScopeInvalid("criteria.version is not a supported SemVer range: %v", err)
			}
			cached = &parsed
			rangeCache[criteria.Version] = cached
		}
		versionRange = cached
	}

	for _, capability := range plugin.Capabilities {
		if capability.Name != criteria.Capability {
			continue
		}
		if versionRange != nil {
			offered, err := ParseVersion(capability.Version)
			if err != nil {
				// Unreachable while registration enforces C-2, so this only
				// fires if a registry invariant was broken. Treating the
				// candidate as non-matching is safer than failing the whole
				// query; the point of the branch is that a malformed version is
				// never silently treated as a *match*.
				continue
			}
			if !versionRange.Match(offered) {
				continue
			}
		}
		if !constraintSetMatches(criteria.Constraints, capability.Constraints) {
			continue
		}
		return true, nil
	}
	return false, nil
}

// Find implements the `find` primitive (§8.1): the plugins visible in scope
// that satisfy criteria.
//
// D-1 requires only plugins inside the current trust scope; the scope is
// resolved through the Core's VisibilityPolicy, so `find` and `watch` cannot
// disagree about visibility (the alternative — filtering inside Find — would let
// a watcher fire for a plugin Find refuses to report).
//
// D-3 is why the result is a plain list and nothing more: discovery proves
// *visibility*, never composability. A plugin returned here still has to pass
// every Bind check, which is why `find` performs none of them.
//
// D-4 allows caching with an invalidation policy. This implementation caches
// nothing: the graph is in memory, a scan is O(plugins), and a cache would need
// an invalidation hook on every mutation — a bug surface with no payoff. The
// per-call SemVer-range cache below is not a discovery cache (it holds parsed
// grammar, not results, and cannot go stale).
//
// D-5 is respected by omission: this function has no side effects on bindings.
func (c *Core) Find(criteria Criteria, scope DiscoveryScope) ([]Identity, error) {
	if err := scope.Valid(); err != nil {
		return nil, err
	}

	c.mu.Lock()
	defer c.mu.Unlock()

	rangeCache := make(map[string]*SemverRange)
	out := make([]Identity, 0, len(c.plugins))
	for _, record := range c.plugins {
		plugin := record.snapshot()
		if !c.visibility.Visible(plugin, scope) {
			continue // D-1
		}
		matched, err := matchesCriteria(plugin, criteria, rangeCache)
		if err != nil {
			return nil, err
		}
		if matched {
			out = append(out, record.identity)
		}
	}
	sort.Slice(out, func(i, j int) bool { return lessIdentity(out[i], out[j]) })
	return out, nil
}

// Watch implements the `watch` primitive (§8.1): it returns a handle and a
// channel of events for plugins that enter the interest set.
//
// The channel is buffered and delivery is non-blocking (see emitLocked), so a
// slow reader degrades to dropped notifications rather than stalling the
// composition operations that produce them. That is consistent with D-3/D-4:
// discovery is a hint, and a watcher that misses an event can always call Find.
//
// The channel is closed by Unwatch and by Reset. A caller MUST NOT assume it
// stays open, and MUST NOT be surprised by a `removed` event it never asked to
// stop — that is exactly what D-6's `removed` type is for.
func (c *Core) Watch(criteria Criteria, scope DiscoveryScope) (int64, <-chan DiscoveryEvent, error) {
	if err := scope.Valid(); err != nil {
		return 0, nil, err
	}

	c.mu.Lock()
	defer c.mu.Unlock()

	c.watchSeq++
	watcher := &watcher{
		id:       c.watchSeq,
		criteria: criteria,
		scope:    scope,
		events:   make(chan DiscoveryEvent, watchBufferSize),
	}
	c.watches[watcher.id] = watcher
	return watcher.id, watcher.events, nil
}

// watchBufferSize bounds a watcher's queue. 64 is far more than any
// control-plane burst observed by the conformance driver; the point of the
// bound is that it exists, so a wedged consumer cannot grow the process without
// limit.
const watchBufferSize = 64

// Unwatch stops a watcher and closes its channel.
//
// Idempotent, like every other operator that removes something: unwatching an
// unknown handle is a no-op, because "that watcher is not running" is already
// true. (Contrast Unbind, where an unknown handle is EAPP_BINDING_INVALID: there
// the caller named a *binding*, and a binding that has never existed is a
// mistake worth reporting. A watcher id is a local subscription handle the
// driver mints, so there is no such thing as naming a nonexistent binding.)
func (c *Core) Unwatch(watchID int64) {
	c.mu.Lock()
	defer c.mu.Unlock()
	watcher, present := c.watches[watchID]
	if !present {
		return
	}
	delete(c.watches, watchID)
	watcher.closed = true
	close(watcher.events)
}

// WatchExists reports whether a watcher handle is still live. The driver uses
// it to answer `discovery.watch` without retaining channels of its own.
func (c *Core) WatchExists(watchID int64) bool {
	c.mu.Lock()
	defer c.mu.Unlock()
	_, present := c.watches[watchID]
	return present
}

// watcher is a live discovery subscription (D-2).
type watcher struct {
	id       int64
	criteria Criteria
	scope    DiscoveryScope
	events   chan DiscoveryEvent
	closed   bool
}

// emitLocked delivers an event to every watcher whose criteria and scope accept
// it (D-2). c.mu MUST be held.
//
// Delivery is non-blocking: a watcher whose buffer is full loses the event
// rather than stalling the mutation that produced it. The Core defines no
// delivery guarantee — that is Delivery semantics, an explicit non-goal (§1.2,
// §20.1) — and blocking here would let a slow subscriber deadlock the operation
// that registers or deactivates a plugin, a far worse failure than a dropped
// notification a watcher can recover from by calling Find.
func (c *Core) emitLocked(event DiscoveryEvent) {
	for _, watcher := range c.watches {
		if watcher.closed {
			continue
		}
		record, present := c.plugins[identityTriple(event.Plugin)]
		if !present {
			continue
		}
		plugin := record.snapshot()
		if !c.visibility.Visible(plugin, watcher.scope) {
			continue // D-2: only events inside the watcher's scope
		}
		rangeCache := make(map[string]*SemverRange)
		matched, err := matchesCriteria(plugin, watcher.criteria, rangeCache)
		if err != nil || !matched {
			continue
		}
		select {
		case watcher.events <- event:
		default:
			// Buffer full: drop. See the doc comment above.
		}
	}
}

File: packages/core/src/discovery.ts
/**
 * Discovery — v3.0.0 §8 (frozen).
 *
 * Invariants: D-1 (find returns only Plugins visible in the current trust scope),
 * D-2 (watch only fires for events inside that scope), D-3 (discovery never implies
 * composability), D-4 (caching is allowed but MUST have an invalidation policy),
 * D-5 (discovery MUST NOT replace binding), D-6 (`DiscoveryEvent.type` is
 * added | removed | changed), D-7 (trust levels are categories, not an ordering).
 *
 * Trust scopes are only honoured when the service actually has a policy for them:
 * a scope the service cannot evaluate is rejected with `EAPP_DISCOVERY_SCOPE_INVALID`
 * instead of being silently ignored. That keeps D-1/D-2 falsifiable.
 */

import type { Constraint } from './capability.js';
import { EappError } from './errors.js';
import type { Identity } from './identity.js';
import { assertValidIdentity } from './identity.js';
import type { Plugin, PluginRef, PluginRegistry } from './plugin.js';
import { satisfiesRange } from './semver.js';

export type TrustLevel = 'L0' | 'L1' | 'L2';

export interface DiscoveryScope {
  trustLevel?: TrustLevel;
  trustDomain?: string;
}

export interface Criteria {
  capability?: string;
  /**
   * A SemVer RANGE, not an exact value — §8.1 declares it as one, and reading it as exact
   * made `find({ version: '^1.0.0' })` silently return nothing. A bare version still means
   * that version exactly; `'*'` means any. See `semver.ts` for the supported grammar.
   */
  version?: string;
  constraints?: Constraint[];
  identity?: Partial<Identity>;
}

export interface DiscoveryEvent {
  type: 'added' | 'removed' | 'changed';
  plugin: PluginRef;
}

/** The two discovery operations (§8.1). */
export interface Discovery {
  find(criteria: Criteria, scope: DiscoveryScope): Promise<PluginRef[]>;
  watch(criteria: Criteria, scope: DiscoveryScope): AsyncIterable<DiscoveryEvent>;
}

/**
 * What the service knows about trust scopes. Only declared levels/domains are
 * accepted; `isVisible` decides visibility inside an accepted scope. There is no
 * implicit hierarchy between levels (D-7).
 */
export interface DiscoveryTrustPolicy {
  trustLevels?: readonly TrustLevel[];
  trustDomains?: readonly string[];
  isVisible?: (plugin: PluginRef, scope: DiscoveryScope) => boolean;
}

/** `watch` result: an async iterable that additionally exposes `close()`. */
export interface DiscoveryWatch extends AsyncIterable<DiscoveryEvent> {
  close(): void;
}

export interface DiscoveryCacheStats {
  hits: number;
  misses: number;
  size: number;
}

export const TRUST_LEVELS: readonly TrustLevel[] = Object.freeze(['L0', 'L1', 'L2'] as const);

export function isTrustLevel(value: unknown): value is TrustLevel {
  return value === 'L0' || value === 'L1' || value === 'L2';
}

function structuralEquals(a: unknown, b: unknown): boolean {
  if (Object.is(a, b)) {
    return true;
  }
  if (typeof a !== 'object' || typeof b !== 'object' || a === null || b === null) {
    return false;
  }
  if (Array.isArray(a) || Array.isArray(b)) {
    if (!Array.isArray(a) || !Array.isArray(b) || a.length !== b.length) {
      return false;
    }
    return a.every((item, index) => structuralEquals(item, b[index]));
  }
  const aRecord = a as Record<string, unknown>;
  const bRecord = b as Record<string, unknown>;
  const aKeys = Object.keys(aRecord);
  if (aKeys.length !== Object.keys(bRecord).length) {
    return false;
  }
  return aKeys.every(
    (key) => Object.hasOwn(bRecord, key) && structuralEquals(aRecord[key], bRecord[key]),
  );
}

function constraintSatisfied(declared: Constraint, wanted: Constraint): boolean {
  return declared.kind === wanted.kind && structuralEquals(declared.value, wanted.value);
}

/** Criteria matching used by both `find` and `watch` (§8.1). */
export function matchesCriteria(plugin: Plugin, criteria: Criteria): boolean {
  const identityFilter = criteria.identity;
  if (identityFilter !== undefined) {
    for (const field of ['domain', 'id', 'instance'] as const) {
      const expected = identityFilter[field];
      if (expected !== undefined && plugin.identity[field] !== expected) {
        return false;
      }
    }
  }

  const name = criteria.capability;
  const version = criteria.version;
  const wanted = criteria.constraints ?? [];
  if (name === undefined && version === undefined && wanted.length === 0) {
    return true;
  }

  return plugin.capabilities.some((capability) => {
    if (name !== undefined && capability.name !== name) {
      return false;
    }
    // §8.1 declares `Criteria.version` a SemVer RANGE, not an exact value. Treating it as
    // exact made `find({ version: '^1.0.0' })` silently return nothing, which is
    // indistinguishable from "no plugin matches".
    if (version !== undefined && !satisfiesRange(capability.version, version)) {
      return false;
    }
    const declared = capability.constraints ?? [];
    return wanted.every((constraint) =>
      declared.some((candidate) => constraintSatisfied(candidate, constraint)),
    );
  });
}

/** Internal unbounded queue backing `watch`; closed queues resolve pending reads. */
class AsyncEventQueue<T> {
  private readonly items: T[] = [];
  private readonly waiters: ((result: IteratorResult<T>) => void)[] = [];
  private closed = false;

  push(item: T): void {
    if (this.closed) {
      return;
    }
    const waiter = this.waiters.shift();
    if (waiter !== undefined) {
      waiter({ value: item, done: false });
      return;
    }
    this.items.push(item);
  }

  next(): Promise<IteratorResult<T>> {
    const item = this.items.shift();
    if (item !== undefined) {
      return Promise.resolve({ value: item, done: false });
    }
    if (this.closed) {
      return Promise.resolve({ value: undefined, done: true });
    }
    return new Promise<IteratorResult<T>>((resolve) => {
      this.waiters.push(resolve);
    });
  }

  close(): void {
    if (this.closed) {
      return;
    }
    this.closed = true;
    this.items.length = 0;
    for (const waiter of this.waiters.splice(0)) {
      waiter({ value: undefined, done: true });
    }
  }

  get size(): number {
    return this.items.length;
  }
}

interface Watcher {
  criteria: Criteria;
  isVisible: (plugin: PluginRef) => boolean;
  queue: AsyncEventQueue<DiscoveryEvent>;
}

interface CacheEntry {
  revision: number;
  refs: readonly PluginRef[];
}

interface ResolvedScope {
  scope: DiscoveryScope;
  isVisible: (plugin: PluginRef) => boolean;
}

export class DiscoveryService implements Discovery {
  private readonly registry: PluginRegistry;
  private readonly policy: DiscoveryTrustPolicy;
  private readonly watchers = new Set<Watcher>();
  private readonly cache = new Map<string, CacheEntry>();
  private readonly unsubscribeRegistry: () => void;
  private hits = 0;
  private misses = 0;

  constructor(registry: PluginRegistry, policy?: DiscoveryTrustPolicy) {
    this.registry = registry;
    this.policy = policy ?? {};
    // D-4 invalidation policy: any registry mutation invalidates cached results.
    this.unsubscribeRegistry = registry.onChange(() => {
      this.invalidate();
    });
  }

  /** D-1: only Plugins visible in the given scope are returned. */
  async find(criteria: Criteria, scope: DiscoveryScope): Promise<PluginRef[]> {
    const resolved = this.resolveScope(scope, 'find');
    const cacheKey = this.cacheKey(criteria, resolved.scope);
    const cached = this.cache.get(cacheKey);
    if (cached !== undefined && cached.revision === this.registry.changeRevision()) {
      this.hits += 1;
      return cached.refs.map((ref) => ref);
    }

    this.misses += 1;
    const refs = this.registry
      .list()
      .filter((plugin) => resolved.isVisible(plugin.identity) && matchesCriteria(plugin, criteria))
      .map((plugin) => plugin.identity);
    const frozen: readonly PluginRef[] = Object.freeze(refs);
    this.cache.set(cacheKey, { revision: this.registry.changeRevision(), refs: frozen });
    return frozen.map((ref) => ref);
  }

  /**
   * D-2: the returned iterable only yields events inside the given scope.
   * The scope gate throws synchronously, before a watcher is registered.
   */
  watch(criteria: Criteria, scope: DiscoveryScope): DiscoveryWatch {
    const resolved = this.resolveScope(scope, 'watch');
    const queue = new AsyncEventQueue<DiscoveryEvent>();
    const watcher: Watcher = {
      criteria: { ...criteria },
      isVisible: resolved.isVisible,
      queue,
    };
    this.watchers.add(watcher);

    let closed = false;
    const close = (): void => {
      if (closed) {
        return;
      }
      closed = true;
      this.watchers.delete(watcher);
      queue.close();
    };

    const iterable: DiscoveryWatch = {
      close,
      [Symbol.asyncIterator](): AsyncIterator<DiscoveryEvent> {
        return (async function* iterate(): AsyncGenerator<DiscoveryEvent> {
          try {
            while (true) {
              const result = await queue.next();
              if (result.done === true) {
                return;
              }
              yield result.value;
            }
          } finally {
            // `for await ... break` calls `return()` on this generator: unregister
            // here so a broken loop does not leak a watcher.
            close();
          }
        })();
      },
    };
    return iterable;
  }

  /** D-6: only added | removed | changed are valid; anything else is rejected. */
  notify(event: DiscoveryEvent): void {
    const source: unknown = event;
    if (typeof source !== 'object' || source === null) {
      throw new EappError('EAPP_DISCOVERY_SCOPE_INVALID', 'DiscoveryEvent MUST be an object');
    }
    const record = source as Record<string, unknown>;
    const type = record['type'];
    if (type !== 'added' && type !== 'removed' && type !== 'changed') {
      throw new EappError(
        'EAPP_DISCOVERY_SCOPE_INVALID',
        `DiscoveryEvent.type MUST be added | removed | changed (D-6), received '${String(type)}'`,
      );
    }
    const plugin = record['plugin'];
    try {
      assertValidIdentity(plugin as Identity);
    } catch (error) {
      throw new EappError('EAPP_DISCOVERY_SCOPE_INVALID', 'DiscoveryEvent.plugin MUST be a valid Identity', {
        details: error,
      });
    }
    const discoveryEvent: DiscoveryEvent = { type, plugin: plugin as PluginRef };

    // Any event invalidates cached results (D-4).
    this.invalidate();

    for (const watcher of [...this.watchers]) {
      if (!watcher.isVisible(discoveryEvent.plugin)) {
        continue; // D-2
      }
      if (!this.matchesEvent(watcher.criteria, discoveryEvent.plugin)) {
        continue;
      }
      watcher.queue.push(discoveryEvent);
    }
  }

  /** Explicit cache invalidation (D-4). */
  invalidate(): void {
    this.cache.clear();
  }

  cacheStats(): DiscoveryCacheStats {
    return { hits: this.hits, misses: this.misses, size: this.cache.size };
  }

  /** Number of live watchers; the async-iterable exit path must return this to 0. */
  watcherCount(): number {
    return this.watchers.size;
  }

  /** Detach from the registry; afterwards the service stops caching/mutating. */
  dispose(): void {
    this.unsubscribeRegistry();
    this.invalidate();
    for (const watcher of [...this.watchers]) {
      this.watchers.delete(watcher);
      watcher.queue.close();
    }
  }

  private matchesEvent(criteria: Criteria, plugin: PluginRef): boolean {
    const registered = this.registry.get(plugin);
    if (registered !== undefined) {
      return matchesCriteria(registered, criteria);
    }
    // A plugin that is no longer registered can only be matched on identity
    // criteria (D-3: discovery never implies composability).
    const hasCapabilityFilter =
      criteria.capability !== undefined ||
      criteria.version !== undefined ||
      (criteria.constraints !== undefined && criteria.constraints.length > 0);
    if (hasCapabilityFilter) {
      return false;
    }
    const detached: Plugin = { identity: plugin, capabilities: [], lifecycle: 'INACTIVE' };
    return matchesCriteria(detached, criteria);
  }

  private cacheKey(criteria: Criteria, scope: DiscoveryScope): string {
    const identityFilter = criteria.identity;
    const identityPart =
      identityFilter === undefined
        ? null
        : [identityFilter.domain ?? null, identityFilter.id ?? null, identityFilter.instance ?? null];
    return JSON.stringify([
      scope.trustLevel ?? null,
      scope.trustDomain ?? null,
      criteria.capability ?? null,
      criteria.version ?? null,
      identityPart,
      criteria.constraints ?? null,
    ]);
  }

  /**
   * Accepts a scope only when the service has a policy for every field it carries.
   * Unknown fields are rejected too: a typo MUST NOT turn D-1/D-2 into a no-op.
   */
  private resolveScope(scope: DiscoveryScope, method: string): ResolvedScope {
    const source: unknown = scope ?? {};
    if (typeof source !== 'object' || source === null || Array.isArray(source)) {
      throw new EappError('EAPP_DISCOVERY_SCOPE_INVALID', `${method}() requires a DiscoveryScope object`);
    }
    const record = source as Record<string, unknown>;
    for (const key of Object.keys(record)) {
      if (key !== 'trustLevel' && key !== 'trustDomain') {
        throw new EappError(
          'EAPP_DISCOVERY_SCOPE_INVALID',
          `Unknown DiscoveryScope field '${key}' (only trustLevel / trustDomain exist)`,
        );
      }
    }

    const trustLevel = record['trustLevel'];
    if (trustLevel !== undefined) {
      if (!isTrustLevel(trustLevel)) {
        throw new EappError('EAPP_DISCOVERY_SCOPE_INVALID', `trustLevel MUST be one of L0 | L1 | L2`);
      }
      if (!(this.policy.trustLevels ?? []).includes(trustLevel)) {
        throw new EappError(
          'EAPP_DISCOVERY_SCOPE_INVALID',
          `No trust policy for level '${trustLevel}': a scope MUST NOT be silently ignored (D-1, D-2)`,
        );
      }
    }

    const trustDomain = record['trustDomain'];
    if (trustDomain !== undefined) {
      if (typeof trustDomain !== 'string' || trustDomain.trim().length === 0) {
        throw new EappError('EAPP_DISCOVERY_SCOPE_INVALID', 'trustDomain MUST be a non-empty string');
      }
      if (!(this.policy.trustDomains ?? []).includes(trustDomain)) {
        throw new EappError(
          'EAPP_DISCOVERY_SCOPE_INVALID',
          `No trust policy for domain '${trustDomain}': a scope MUST NOT be silently ignored (D-1, D-2)`,
        );
      }
    }

    const resolvedScope: DiscoveryScope = {
      ...(trustLevel !== undefined ? { trustLevel } : {}),
      ...(trustDomain !== undefined ? { trustDomain } : {}),
    };
    const predicate = this.policy.isVisible;
    const isVisible =
      predicate === undefined
        ? (): boolean => true
        : (plugin: PluginRef): boolean => predicate(plugin, resolvedScope);
    return { scope: resolvedScope, isVisible };
  }
}
